lib/dispatcher/agent.js

```diff
@@ -39,34 +39,71 @@ class Agent extends DispatcherBase {
     this[kOptions] = { ...util.deepClone(options), connect }
     this[kFactory] = factory
     this[kClients] = new Map()
+    this[kQueue] = []
+
+    if (options.maxGlobalConnections != null) {
+      if (!Number.isFinite(options.maxGlobalConnections) || options.maxGlobalConnections <= 0) {
+        throw new InvalidArgumentError('maxGlobalConnections must be a number greater than 0')
+      }
+
+      if (options.connections != null && options.maxGlobalConnections < options.connections) {
+        throw new InvalidArgumentError('maxGlobalConnections must be greater than or equal to connections')
+      }
+    }
+
+    this[kOptions].maxGlobalConnectionsReached = () => {
+      if (options.maxGlobalConnections == null) return false
+      return (this[kConnected] >= options.maxGlobalConnections)
+    }
 
     this[kOnDrain] = (origin, targets) => {
-      this.emit('drain', origin, [this, ...targets])
+      this.emit('drain', origin.origin, [this, ...targets])
     }
 
     this[kOnConnect] = (origin, targets) => {
-      const result = this[kClients].get(origin)
+      const key = origin.origin
+      const result = this[kClients].get(key)
       if (result) {
         result.count += 1
       }
-      this.emit('connect', origin, [this, ...targets])
+      this.emit('connect', key, [this, ...targets])
     }
 
     this[kOnDisconnect] = (origin, targets, err) => {
-      const result = this[kClients].get(origin)
+      const key = origin.origin
+      const result = this[kClients].get(key)
       if (result) {
         result.count -= 1
-        if (result.count <= 0) {
-          this[kClients].delete(origin)
+        if (result.count <= 0 && err?.code !== 'UND_ERR_INFO') {
+          this[kClients].delete(key)
           result.dispatcher.destroy()
         }
       }
-      this.emit('disconnect', origin, [this, ...targets], err)
+      this.emit('disconnect', key, [this, ...targets], err)
+      this[kResume]()
     }
 
     this[kOnConnectionError] = (origin, targets, err) => {
       // TODO: should this decrement result.count here?
-      this.emit('connectionError', origin, [this, ...targets], err)
+      this.emit('connectionError', origin.origin, [this, ...targets], err)
+    }
+  }
+
+  get [kConnected] () {
+    return this[kClients].values().reduce((acc, { count = 0 }) => acc + count, 0)
```

Comment:
try avoiding a reduce(), it's super slow. Also the continuous conversion of `kClients` to values will slow things down significantly. We probably need to maintain two data structures for them.